tests/cpp_unit_tests/test_tap_position_optimizer.cpp

```diff
@@ -2,67 +2,303 @@
+        }
+
+        const pgm_tap::WeightedTrafoList all_edge_weights = get_edge_weights(g);
+        const pgm_tap::WeightedTrafoList ref_edge_weights{{true, {0, 1}, 1}, {true, {0, 2}, 2}, {true, {2, 3}, 5}};
         CHECK(all_edge_weights == ref_edge_weights);
     }
 
     SUBCASE("Sorting transformer edges") {
-        pgm_tap::WeightedTrafoList trafoList{{Idx2D{1, 1}, pgm_tap::infty},
-                                             {Idx2D{1, 2}, 5},
-                                             {Idx2D{1, 3}, 4},
-                                             {Idx2D{2, 1}, 4},
-                                             {Idx2D{2, 2}, 3},
-                                             {Idx2D{3, 1}, 2},
-                                             {Idx2D{3, 2}, 1},
-                                             {Idx2D{3, 3}, 1}};
-
-        const pgm_tap::RankedTransformerGroups referenceList{{Idx2D{3, 2}, Idx2D{3, 3}}, {Idx2D{3, 1}}, {Idx2D{2, 2}},
-                                                             {Idx2D{1, 3}, Idx2D{2, 1}}, {Idx2D{1, 2}}, {Idx2D{1, 1}}};
-
-        auto const sortedTrafoList = pgm_tap::rank_transformers(trafoList);
+        pgm_tap::WeightedTrafoList trafoList{{false, Idx2D{1, 1}, pgm_tap::infty},
+                                             {true, Idx2D{1, 2}, 5},
+                                             {true, Idx2D{1, 3}, 4},
+                                             {true, Idx2D{2, 1}, 4},
+                                             {true, Idx2D{2, 2}, 3},
+                                             {true, Idx2D{3, 1}, 2},
+                                             {true, Idx2D{3, 2}, 1},
+                                             {true, Idx2D{3, 3}, 1}};
+
+        const pgm_tap::RankedTransformerGroups referenceList{
+            {Idx2D{3, 2}, Idx2D{3, 3}}, {Idx2D{3, 1}}, {Idx2D{2, 2}}, {Idx2D{1, 3}, Idx2D{2, 1}}, {Idx2D{1, 2}}};
+
+        const pgm_tap::RankedTransformerGroups sortedTrafoList = pgm_tap::rank_transformers(trafoList);
         CHECK(sortedTrafoList == referenceList);
     }
+
+    SUBCASE("Ranking complete the graph") {
+        pgm_tap::RankedTransformerGroups order = pgm_tap::rank_transformers(state);
+        pgm_tap::RankedTransformerGroups ref_order{{Idx2D{0, 1}, Idx2D{0, 1}, Idx2D{0, 4}, Idx2D{0, 5}},
+                                                   {Idx2D{2, 3}, Idx2D{4, 5}, Idx2D{5, 4}, Idx2D{5, 7}},
+                                                   {Idx2D{8, 9}}};
+        // necessary evil, made possible by platform differences
+        auto compareIdx2D = [](const Idx2D& a, const Idx2D& b) {
+            if (a.group != b.group) {
+                return a.group < b.group;
+            }
+            if (a.pos != b.pos) {
+                return a.pos < b.pos;
+            }
+            if (a.group != b.group) {
+                return a.group < b.group;
+            }
+            return a.pos < b.pos;
+        };
```

Comment:
???????

1. there's duplication
2. since `Idx2D` represents a hierarchical index, it's probably good to implement this function where it's declared anyways. this is not the correct location.
3. even more so, it is probably possible to `default` it cause by default it is strong ordering in the order the fields are declared
4. if you really need it here, please make it like this
```suggestion
        // necessary evil, made possible by platform differences
        auto compareIdx2D = [](const Idx2D& a, const Idx2D& b) {
            if (a.group != b.group) {
                return a.group < b.group;
            }
            return a.pos < b.pos;
        };
```

i'm very interested to see what the platform differences are

Reply:
> ???????

?????????

> 
> 1. there's duplication

Don't work on a Sunday afternoon.

> 2. since `Idx2D` represents a hierarchical index, it's probably good to implement this function where it's declared anyways. this is not the correct location.
> 3. even more so, it is probably possible to `default` it cause by default it is strong ordering in the order the fields are declared
> 4. if you really need it here, please make it like this
> 

As mentioned in the code, only necessary for the test cases.

> i'm very interested to see what the platform differences are

{..., {5, 7}, {5, 4}, ...}
vs
{..., {5, 4}, {5, 7}, ...}

Reply:
let's see if the same thing happens if you declare the strong ordering where `Idx2D` is implemented. IMO it makes sense to add fundamental behavior like this on a global level (reinforcing the notion of hierarchical indices)

Reply:
how, when and why do these differences arise? is it caused by boost?

as long as the weak ordering criterion is satisfied, that's OK and that's exactly what you check